deprecate and remove Kubelet RunOnce mode (#4581)
kubernetes/enhancements: keps/sig-node/4580-deprecate-kubelet-runonce/README.md

- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Deprecated and removed kubelet support for RunOnce mode.

## Motivation

> Runonce mode doesn't support many newer pod features (init containers), and the pod lifecycle for runonce mode is even less well defined than normal.

Reviewer: This needs a bit more context. The feature is also not tested.

### Goals

* Mark the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet as deprecated.
* Remove kubelet support for RunOnce mode.

### Non-Goals

> Runonce mode will be removed immediately. The deprecation and removal process will be gradual and feature gate to increase awareness among potential users.

Reviewer: You mean Runounce mode will **not** be removed immediately? These sentences conflict.

Author: This is a stupid typo.

Author: Here is the Non-Goals part, we will not remove the Runonce mode immediately, because the deprecation and removal process will be gradual.....

## Proposal

This proposal outlines plans to deprecate and remove runonce mode in kubelet.

> ### Risks and Mitigations

Reviewer: The risk is that someone is still relying on this featuer. This is a good place to mention podman as a viable alternative.

## Design Details

### KubeletConfiguration Change: KubeletConfiguration

Mark the `RunOnce` field as Deprecated.

### kubelet flag Change

make the `runonce` flag as Deprecated.

### Implement warning logging for runonce mode usage

Starting in 1.31, during kubelet startup, if running in runonce mode, the kubelet will log a warning message, for example:

```

> klog.Warning("runonce has been deprecated")

Reviewer: ```suggestion
klog.Warning("RunOnce mode has been deprecated, and will be removed in a future release")
```

```

### Introduction RemoveRunonce feature gate

With the introduction of the `RemoveRunonce` feature gate, Kubernetes aims to guide users through the deprecated runonce mode. When this feature gate is enabled, kubelet will refuse to start when the runonce flag is set.

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

> ##### Prerequisite testing updates

Reviewer 2: Are there any e2e tests we need to remove as part of this deprecation?

Author: runonce mode doesn't have any e2e tests, only some unit tests.

- Feature gate `RemoveRunonce` is introduced, is disabled by default. Enabling this feature gate will fail the kubelet on startup with runonce mode enable.

#### Beta

- `RemoveRunonce` feature gate is enabled by default.
- Failed when starting kubelet in runonce mode.

#### GA

- Remove `RemoveRunonce`  feature gate.
- Remove the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet
- Remove kubelet runonce mode.

### Upgrade / Downgrade Strategy

### Version Skew Strategy

- N/A

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: RemoveRunonce
  - Components depending on the feature gate: kubelet
  
  - Will enabling / disabling the feature require downtime of the control
    plane? Yes. Flag must be set on kubelet start. To disable, kubelet must be restarted. Hence, there would be brief control component downtime on a given node.
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? Yes. See above; disabling would require brief node downtime.

###### Does enabling the feature change any default behavior?

> No.

Reviewer 2: ```suggestion
Yes, this feature will fail kubelet if runonce is enabled.
```

Reviewer 2: Eh. The feature will change the default behavior of runonce but not sure if that is what this question is asking actually..

Author: We did not change the "default behavior" because the default value for runonce mode is `false`.

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Deprecated and removed kubelet support for RunOnce mode.

## Motivation

* Runonce does not work in systemd mode.
* Runonce mode doesn't support many newer pod features (init containers)
* The pod lifecycle for runonce mode is even less well defined than normal.
* The mode appears to be entirely untested in CI, except for a few unit tests.

### Goals

> * Mark the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet as deprecated.

Reviewer 3: Do we intend to leave the `--runonce` command line argument as part of the kubelet indefinitelty? If not, let's be clear.

For the configuration file, we may have to leave it in as a field where only one value is actually supported.

* Remove kubelet support for RunOnce mode.

### Non-Goals

Runonce mode will be removed immediately. The deprecation and removal process will be gradual and feature gate to increase awareness among potential users.

> 

Reviewer: It will probably be helpful to add a brief background section. What is runonce mode?

## Proposal

This proposal outlines plans to deprecate and remove runonce mode in kubelet.

### Risks and Mitigations

> Some people still rely on this feature, but podman addresses the same use case with more well-supported way, ref: https://docs.podman.io/en/latest/markdown/podman-kube.1.html. Affected users can migrate to *podman kube subcommand* on demand.

Reviewer: ```suggestion
Some people may still rely on this feature, but podman addresses the same use case with more well-supported way, ref: https://docs.podman.io/en/latest/markdown/podman-kube.1.html. Affected users can migrate to *podman kube subcommand* on demand.
```

I don't think we actually know of any users (unless you do?). Maybe someone will come forth during the review process. Otherwise, that's the purpose of the feature-gated deprecation rollout.

## Design Details

### KubeletConfiguration Change: KubeletConfiguration

Mark the `RunOnce` field as Deprecated.

### kubelet flag Change

> make the `runonce` flag as Deprecated.

Reviewer 3: Why not remove the command line argument too? (eventually)

### Implement warning logging for runonce mode usage

Starting in 1.31, during kubelet startup, if running in runonce mode, the kubelet will log a warning message, for example:

```
klog.Warning("RunOnce mode has been deprecated, and will be removed in a future release")
```

### Introduction RemoveRunonce feature gate

> With the introduction of the `RemoveRunonce` feature gate, Kubernetes aims to guide users through the deprecated runonce mode. When this feature gate is enabled, kubelet will refuse to start when the runonce flag is set.

Reviewer 3: Name the gate after the feature, not the change you intend to make. For example, call it `KubeletRunOnceMode`.

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

- N/A

##### Integration tests

- N/A

##### e2e tests

- N/A

### Graduation Criteria

#### Alpha

> - Feature gate `RemoveRunonce` is introduced, is disabled by default. Enabling this feature gate will fail the kubelet on startup with runonce mode enable.

Reviewer 3: Name the gate after the feature, not the change you intend to make. For example, call it `KubeletRunOnceMode`.

### Version Skew Strategy

- N/A

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: RemoveRunonce
  - Components depending on the feature gate: kubelet
  
  - Will enabling / disabling the feature require downtime of the control
    plane? Yes. Flag must be set on kubelet start. To disable, kubelet must be restarted. Hence, there would be brief control component downtime on a given node.
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? Yes. See above; disabling would require brief node downtime.

###### Does enabling the feature change any default behavior?

No.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Using the feature gate is the only way to enable/disable this feature.

###### What happens if we reenable the feature if it was previously rolled back?

N/A

###### Are there any tests for feature enablement/disablement?

N/A

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

> In the alpha `stage`, this feature is disabled by default.
> 
> In the beta `stage`, this feature is enabled by default. With this feature enabled, the kubelet will refuse to start if is still using runonce mode.

Reviewer 3: ```suggestion
For the first release that deprecates the mode, the `RunOnce` mode is deprecated but available by default.
Cluster operators can test the behavior by disabling the feature gate.

For a later release following deprecation, the `RunOnce` mode is deprecated and disabled by default.
Cluster operators can reinstate the mode by explicitly enabling the feature gate.
```

###### Does this feature depend on any specific services running in the cluster?

### Scalability

###### Will enabling / using this feature result in any new API calls?

No

###### Will enabling / using this feature result in introducing new API types?

No

###### Will enabling / using this feature result in any new calls to the cloud provider?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

###### What are other known failure modes?

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

- \- 2024-04-17: Initial draft KEP

## Drawbacks

## Alternatives

> 

Reviewer 3: > * Runonce does not work in systemd mode.

Let's document the alternative of fixing this (systemd supports one shot units).

Remember, documenting an alternative is not the same as **selecting** an alternative. We document alternatives to show that we considered them, but they didn't get selected as our preferred approach.

Reviewer 4: This argument holds.

Reviewer 4: This argument holds.

Author: Added.

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> Deprecated and remove kubelet support for RunOnce mode, and mark the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet as deprecated,  and finally remove the `runonce` flag.

Reviewer 4: ```suggestion
Deprecate and remove kubelet support for RunOnce mode, and mark the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet as deprecated, finally remove the `runonce` flag.
```

- Comment the `RunOnce` field in KubeletConfiguration as 'no longer has any effect', and remove the kubelet's `runonce` flag.
- Remove kubelet Runonce mode.

### Upgrade / Downgrade Strategy

### Version Skew Strategy

- N/A

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: LegacyNodeRunOnceMode
  - Components depending on the feature gate: kubelet
  - Will enabling / disabling the feature require downtime of the control
    plane? Yes. Flag must be set on kubelet start. To disable, kubelet must be restarted. Hence, there would be brief control component downtime on a given node.
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? Yes. See above; disabling would require brief node downtime.

###### Does enabling the feature change any default behavior?

No.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Using the feature gate is the only way to enable/disable this feature.

###### What happens if we reenable the feature if it was previously rolled back?

> N/A

Reviewer 4: ```suggestion
Re-enabling the feature will make the RunOnce functionality available in the kubelet. 
```

###### Are there any tests for feature enablement/disablement?

N/A

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

In the alpha `stage`, this feature is enable by default.

Cluster operators can test the behavior by enabling the feature gate.

> In the beta `stage`, this feature is enable by default. With this feature disable, the kubelet will refuse to start if is still using Runonce mode.

Reviewer 4: ```suggestion
In the beta `stage`, this feature is enable by default. With this feature disabled, the kubelet will refuse to start if is still using Runonce mode.
```

> Cluster operators can reinstate the mode by explicitly enabling the feature gate.

Reviewer 4: Is there a metric or a way for a cluster administrator to verify if this functionality is being used other than checking the kubelet config file and invocation?

Author: No, there is no other way for cluster administrator to verify whether this feature is enabled.

Reviewer 3: We _could_ have kubelets in RunOnce mode emit a special (alpha?) metric, and delay the switch to off-by-default until all supported releases do emit that metric. If we want.

Author: Considering that this KEP is to deprecate a feature, I am not sure if we have enough motivation to add a new metric for it. Do we have similar implementations in the past?

Reviewer 4: No, you're right we can get away without adding one.

###### What specific metrics should inform a rollback?

N/A

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

N/A

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

> N/A

Reviewer 4: You are removing flags and configuration fields, so you need to fill this one in.

### Monitoring Requirements

* N/A

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

> 

Reviewer 4: No.

- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
  - [ ] e2e Tests for all Beta API Operations (endpoints)
  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]

> - [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

Reviewer 4: Nit: make sure to check the appropriate items in this list.

Reviewer 4: This nit still holds.

Author: Sorry for my carelessness, it has been fixed.

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> Deprecated and removed kubelet support for RunOnce mode.

Reviewer 4: ```suggestion
Deprecate and remove kubelet support for RunOnce mode.
```

## Motivation

* Runonce does not work in systemd mode.

> * Runonce mode doesn't support many newer pod features (init containers).

Reviewer 4: Nit, but it would be nice to stick with `RunOnce` nomenclature consistently throughout the doc.

Reviewer 4: This still holds.

Author: Sorry for my carelessness, it has been fixed.

> * The pod lifecycle for runonce mode is even less well defined than normal.

Reviewer 4: Not sure what you mean here by less defined?

> * The mode appears to be entirely untested in CI, except for a few unit tests.

Reviewer 4: It's either *entirely* untested in CI, or has a few unit tests, which we run ci, afaict :wink:

### Goals

> * Mark the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet as deprecated,  and finally remove the runonce flag.

Reviewer 4: Can this statement be also the summary? Ref https://youtu.be/JZ9LQR_j0Rk?t=1482

* Remove kubelet support for RunOnce mode.

### Non-Goals

Immediate removal: the deprecation and removal process will be gradual and feature gate to increase awareness among potential users.

## Proposal

The runonce mode of kubelet will exit the kubelet process after spawning pods from the local manifests or remote URL. It is suitable for scenarios where one-time tasks need to be run on the node, this proposal outlines plans to deprecate and remove runonce mode in kubelet.

### Risks and Mitigations

> Some people may still rely on this feature, but podman addresses the same use case with more well-supported way, ref: https://docs.podman.io/en/latest/markdown/podman-kube.1.html. Affected users can migrate to *podman kube subcommand* on demand.

Reviewer 4: What about users of docker or other runtimes?

Reviewer 4: Do you have at least a rough estimate how many users this might affect?

Author: > Do you have at least a rough estimate how many users this might affect?

I can't estimate how many users this will affect, and in fact a Google search yields very few blogs or discussions about the Runonce model.

> What about users of docker or other runtimes?

Maybe they can install a podman, podman and docker do not conflict.

Reviewer 4: Let's make sure we call out that affected users have alternatives mentioning both podman and docker.

Author: Because Docker does not support this mode and there is no reliable third-party tool available, I added a description of manually translating Kubernetes YAML to Docker Compose files. Does this meet our expectations?

## Design Details

### KubeletConfiguration Change: KubeletConfiguration

Mark the `RunOnce` field as deprecated.

### kubelet flag Change

make the `runonce` flag as deprecated, and remove it in GA version.

### Implement warning logging for runonce mode usage

Starting in 1.31, during kubelet startup, if running in runonce mode, the kubelet will log a warning message, for example:

```
klog.Warning("RunOnce mode has been deprecated, and will be removed in a future release")
```

### Introduction LegacyNodeRunOnceMode feature gate

> With the introduction of the `LegacyNodeRunOnceMode` feature gate, Kubernetes aims to guide users through the deprecated runonce mode. When this feature gate is enabled, kubelet will refuse to start when the runonce flag is set.

Reviewer 4: `LegacyNodeRunOnceMode` suggests the opposite of what you're proposing here. Iow. setting `LegacyNodeRunOnceMode=true` it will not allow using the deprecated RunOnce mode. Maybe something like `RemovedNodeRunOnceMode` or something that explicitly says it's not there will be a better feature gate name?

Author: The reason for using Legacy* is that we have used it in some areas...
ref: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates-removed/

Reviewer 3: We can invert the sense of the feature gate: ie, you disable it to test running without the deprecated mode, and the feature gate turns false-by-default before we eventually remove it.

Reviewer 3: Try:
```diff
-Once this feature gate is enabled, kubelet will refuse to start when the runonce flag is set.
+Unless this feature gate is enabled, kubelet will refuse to start when the `--runonce` command line flag is set.
```

Author: I reversed it wherever it was mentioned (hopefully this won't get any worse).

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

- N/A

##### Integration tests

- N/A

##### e2e tests

- N/A

### Graduation Criteria

#### Alpha

- Feature gate `LegacyNodeRunOnceMode` is introduced, is disabled by default. Enabling this feature gate will fail the kubelet on startup with runonce mode enable.
- Mark the `RunOnce` field in KubeletConfiguration as deprecated.

#### Beta

- `LegacyNodeRunOnceMode` feature gate is enabled by default.
- Failed when starting kubelet in runonce mode.

#### GA

> - Remove `LegacyNodeRunOnceMode`  feature gate.

Reviewer 4: The general flow is that at GA we make the feature on by default, w/o the ability to be turned off. Only GA+2 is when we drop a feature from https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go. You can check that file for examples.

Reviewer 3: Sure, but for **this** the flow will be deprecated (but active by default), to still deprecated (but you need to opt in), to removed.

This is a feature _deprecation_.

Reviewer 3: Also see resolved comment https://github.com/kubernetes/enhancements/pull/4581/files#r1592808502

Reviewer 4: Yeah, with the inverted approach it'll be fine.

- Comment the `RunOnce` field in KubeletConfiguration as 'no longer has any effect', and remove the kubelet's runonce flag.
- Remove kubelet runonce mode.

### Upgrade / Downgrade Strategy

### Version Skew Strategy

- N/A

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

- [x] Feature gate (also fill in values in `kep.yaml`)

>   - Feature gate name: LegacyNodeRunOnceMode

Reviewer 4: See my other comment about feature gate name.

Reviewer 3: We might need a different template for this kind of change.
